# Introduction
some words about Bangalore, task, EDA

# Libraries

In [57]:
# analysis
import numpy as np 
import pandas as pd 

# visualization
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

# 1. Data Overview

In [76]:
data = pd.read_csv('/kaggle/input/zomato-bangalore-restaurants/zomato.csv')
data.sample(5)

url  \
42766  https://www.zomato.com/bangalore/babayagas-ind...   
17996  https://www.zomato.com/bangalore/tayabbs-indir...   
18770  https://www.zomato.com/bangalore/jimis-beer-ca...   
10675  https://www.zomato.com/bangalore/al-habibi-res...   
3877   https://www.zomato.com/bangalore/punjabi-tadka...   

                                                 address  \
42766  949, 3rd Floor, 12th Main Road, HAL 2nd Stage,...   
17996  46, Opposite BSNL Office, 80 Feet Road, Indira...   
18770  1081, 12th Main, HAL 2nd Stage, Off 100 Feet R...   
10675    Narayan Gowda Layout, 1st Stage, BTM, Bangalore   
3877   2733, 16th Cross, 27th Main, Opposite NIFT, Se...   

                       name online_order book_table    rate  votes  \
42766            BabaYaga's           No        Yes  4.0 /5    241   
17996               Tayabbs          Yes         No   4.0/5    736   
18770      Jimi's Beer Cafe           No        Yes   3.9/5    815   
10675  Al Habibi Restaurant          Yes         No   3.5/5     17   
3877          Punjabi Tadka          Yes         No   2.8/5    188   

                                  phone     location           rest_type  \
42766                      080 43756555  Indiranagar  Bar, Casual Dining   
17996                    +91 7829811234  Indiranagar         Quick Bites   
18770    +91 8892607344\r\n080 25203879  Indiranagar                 Bar   
10675                    +91 9611118091          BTM         Quick Bites   
3877   +91 9611765699\r\n+91 9980845823          HSR       Casual Dining   

                                              dish_liked  \
42766  Masala Fries, Cocktails, Chicken Wings, Beer, ...   
17996  Rolls, Beda Roti, Chicken Malai Roll, Peri Per...   
18770  Beer, Beef Steak, Potato Wedges, Burgers, Nach...   
10675                                                NaN   
3877   Paneer Parantha, Mutton Biryani, Lassi, Dal Ma...   

                              cuisines approx_cost(for two people)  \
42766            American, Finger Food                       1,500   
17996   Rolls, North Indian, Fast Food                         500   
18770  American, North Indian, Chinese                       1,700   
10675            North Indian, Chinese                         300   
3877             North Indian, Chinese                         400   

                                            reviews_list  \
42766  [('Rated 2.0', "RATED\n  It was one of the wor...   
17996  [('Rated 2.0', 'RATED\n  Ordered double mutton...   
18770  [('Rated 4.0', "RATED\n  This place is a great...   
10675  [('Rated 1.0', 'RATED\n  First chowmein where ...   
3877   [('Rated 3.0', 'RATED\n  The food was well pac...   

                                               menu_item listed_in(type)  \
42766                                                 []        Delivery   
17996                                                 []        Delivery   
18770                                                 []        Dine-out   
10675  ['Jeera Rice', 'Paneer Butter Masala', 'Bonele...        Dine-out   
3877                                                  []        Delivery   

      listed_in(city)  
42766         MG Road  
17996     Indiranagar  
18770     Indiranagar  
10675             BTM  
3877        Bellandur

In [20]:
print(f'number of entries: {len(data)}')
print(f'number of features: {len(data.columns)}')

number of entries: 51717
number of features: 17


In [29]:
pd.concat([data.dtypes, data.isna().sum(), data.nunique()], axis=1).rename(columns={0:'column type', 1:'number of NaN', 2:'number of unique'})

column type  number of NaN  number of unique
url                              object              0             51717
address                          object              0             11495
name                             object              0              8792
online_order                     object              0                 2
book_table                       object              0                 2
rate                             object           7775                64
votes                             int64              0              2328
phone                            object           1208             14926
location                         object             21                93
rest_type                        object            227                93
dish_liked                       object          28078              5271
cuisines                         object             45              2723
approx_cost(for two people)      object            346                70
reviews_list                     object              0             22513
menu_item                        object              0              9098
listed_in(type)                  object              0                 7
listed_in(city)                  object              0                30

description of features + first observation (about missing values and strange types)
<>

**Columns description:**

* url contains the url of the restaurant in the zomato website

* address contains the address of the restaurant in Bengaluru

* name contains the name of the restaurant

* online_order whether online ordering is available in the restaurant or not

* book_table table book option available or not

* rate contains the overall rating of the restaurant out of 5

* votes contains total number of rating for the restaurant as of the above mentioned date

* phone contains the phone number of the restaurant

* location contains the neighborhood in which the restaurant is located

* rest_type restaurant type

* dish_liked dishes people liked in the restaurant

* cuisines food styles, separated by comma

* approx_cost(for two people) contains the approximate cost for meal for two people

* reviews_list list of tuples containing reviews for the restaurant, each tuple

* menu_item contains list of menus available in the restaurant

* listed_in(type) type of meal

* listed_in(city) contains the neighborhood in which the restaurant is listed

## 1.2 look closely to some column

In [38]:
data.name.value_counts()

Cafe Coffee Day                                            96
Onesta                                                     85
Just Bake                                                  73
Empire Restaurant                                          71
Five Star Chicken                                          70
                                                           ..
Natis                                                       1
Shreyas                                                     1
Vishwas Kabab Centre                                        1
Taj Biryani Centre                                          1
Plunge - Sheraton Grand Bengaluru Whitefield Hotel &...     1
Name: name, Length: 8792, dtype: int64

This is restaurant's chains so this is only 8792

In [69]:
data.rate.unique()

array(['4.1/5', '3.8/5', '3.7/5', '3.6/5', '4.6/5', '4.0/5', '4.2/5',
       '3.9/5', '3.1/5', '3.0/5', '3.2/5', '3.3/5', '2.8/5', '4.4/5',
       '4.3/5', 'NEW', '2.9/5', '3.5/5', nan, '2.6/5', '3.8 /5', '3.4/5',
       '4.5/5', '2.5/5', '2.7/5', '4.7/5', '2.4/5', '2.2/5', '2.3/5',
       '3.4 /5', '-', '3.6 /5', '4.8/5', '3.9 /5', '4.2 /5', '4.0 /5',
       '4.1 /5', '3.7 /5', '3.1 /5', '2.9 /5', '3.3 /5', '2.8 /5',
       '3.5 /5', '2.7 /5', '2.5 /5', '3.2 /5', '2.6 /5', '4.5 /5',
       '4.3 /5', '4.4 /5', '4.9/5', '2.1/5', '2.0/5', '1.8/5', '4.6 /5',
       '4.9 /5', '3.0 /5', '4.8 /5', '2.3 /5', '4.7 /5', '2.4 /5',
       '2.1 /5', '2.2 /5', '2.0 /5', '1.8 /5'], dtype=object)

some words about neccesessary to correct

In [67]:
data[["location", "listed_in(city)"]]

location listed_in(city)
0                    Banashankari    Banashankari
1                    Banashankari    Banashankari
2                    Banashankari    Banashankari
3                    Banashankari    Banashankari
4                    Basavanagudi    Banashankari
...                           ...             ...
51712                  Whitefield      Whitefield
51713                  Whitefield      Whitefield
51714                  Whitefield      Whitefield
51715  ITPL Main Road, Whitefield      Whitefield
51716  ITPL Main Road, Whitefield      Whitefield

[51717 rows x 2 columns]

we could see that descriptions of these columns are the same

In [ ]:
data[data['location'] != data['listed_in(city)']][['location', 'listed_in(city)']]

# 2. Data cleaning

In [24]:
# check duplicates
data[data.duplicated()]

Empty DataFrame
Columns: [url, address, name, online_order, book_table, rate, votes, phone, location, rest_type, dish_liked, cuisines, approx_cost(for two people), reviews_list, menu_item, listed_in(type), listed_in(city)]
Index: []

In [28]:
# drop columns which is less informative
data.drop(['url', 'address', 'phone'], inplace=True)

(8792, 51717, 11495)

In [78]:
# correct 'rate' column
data.rate = data.rate.replace({'-': np.nan, 'NEW': np.nan})
data.rate = data.rate.astype(str)
data.rate = data.rate.apply(lambda row: row.replace('/5', ''))
data.rate.astype(float)

AttributeError: 'StringMethods' object has no attribute 'apply'

In [ ]:
# drop column which has a lot of NaN values
data.drop(['dish_liked'], inplace=True)


# drop entries with NaN values (for EDA; after that we can return it and use imputer to fill)
data.drop()